File: 3244-shortest-distance-after-road-addition-queries-ii/3244-shortest-distance-after-road-addition-queries-ii.java
class Solution {

    public int[] shortestDistanceAfterQueries(int n, int[][] queries) {
        int[] result = new int[queries.length];
        TreeMap<Integer, Integer> map = new TreeMap<>();
        for (int i = 0, k = n - 1; i < queries.length; i++) {
            Map.Entry<Integer, Integer> entry = map.floorEntry(queries[i][0]);
            if (entry == null || entry.getValue() < queries[i][1]) {
                for (; (entry = map.ceilingEntry(queries[i][0])) != null && entry.getValue() <= queries[i][1]; map.remove(entry.getKey())) {
                    k += entry.getValue() - entry.getKey() - 1;
                }
                map.put(queries[i][0], queries[i][1]);
                k -= queries[i][1] - queries[i][0] - 1;
            }
            result[i] = k;
        }
        return result;
    }
}

File: 3244-shortest-distance-after-road-addition-queries-ii/3244-shortest-distance-after-road-addition-queries-ii.cpp
#include <bits/stdc++.h>
using namespace std;
#define SZ 1333333
#define fi first
#define se second
typedef long long ll;
#define mp make_pair
#define pb push_back
#define pii pair<int,int>

class Solution {
public:
    vector<int> shortestDistanceAfterQueries(int n, vector<vector<int>>& queries) {
        int ans=0;
        map<int,int> pr;
        vector<int> aa;
        for(auto q:queries) {
            int a=q[0],b=q[1];
            auto s=pr.upper_bound(a);
            if(s!=pr.begin()) {
                auto u=--s;
                // check if (a,b) is covered
                if(u->se>=b) goto skip;
            }
            // remove all the covered
            s=pr.lower_bound(a);
            while(s!=pr.end()&&s->se<=b)
                ans-=s->se-s->fi-1,
                s=pr.erase(s);
            pr[a]=b;
            ans+=b-a-1;
            skip:;
            aa.pb(n-1-ans);
        }
        return aa;
    }
};